File: posts/5.a-more-type-driven-erlang-workflow-maybe/erlang_types/src/phone_service.erl
-module(phone_service).

-export([call/1]).

-spec call(contact_information:t()) -> integer().
call(Contact) ->
    %% unpack the contact info
    URL = <<"http://example.com/call/", (maps:get(phone_number, Contact))/bytes>>,
    {ok, Status} = post(URL),
    Status.

-spec post(binary()) -> {ok, integer()}.
post(URL) when is_binary(URL) ->
    %% some external call happens here
    {ok, 200}.

File: posts/x.a-more-type-driven-erlang-workflow-maybe/elixir_types/lib/phone_service.ex
defmodule PhoneService do

  @spec call(ContactInformation.t()) :: integer
  def call(contact) do
    url = "http://example.com/call/" <> contact.phone_number
    {:ok, status} = post(url)
    status
  end

  defp post(url) when is_binary(url) do
    # some external call happens here
    {:ok, 200}
  end
end
